runtime/src/main/java/org/corfudb/runtime/view/ManagementView.java

```diff
@@ -320,125 +238,303 @@ private NodeStatus getLogUnitNodeStatusInLayout(Layout layout, String server) {
-            return new ClusterStatusReport(null,
-                    layoutServers.stream().collect(Collectors.toMap(
-                            endpoint -> endpoint,
-                            node -> NodeStatus.DOWN)),
-                    ClusterStatus.UNAVAILABLE);
+    private Map<String, Layout> getLayouts(List<String> layoutServers, Boolean discoverLayoutServers) {
+        Map<String, CompletableFuture<Layout>> layoutFuturesMap = getLayoutFutures(layoutServers);
+        Map<String, Layout> layoutMap = new HashMap<>();
+
+        // Wait on the Completable futures
+        for (Entry<String, CompletableFuture<Layout>> entry : layoutFuturesMap.entrySet()) {
+            try {
+                Layout nodeLayout = entry.getValue().get();
+                log.debug("Server:{} responded with layout: {}", entry.getKey(), nodeLayout);
+                layoutMap.put(entry.getKey(), nodeLayout);
+            } catch (Exception e) {
+                log.error("getClusterStatus: Error while fetching layout from {}. Exception: {}",
+                        entry.getKey(), e);
+            }
         }
 
-        RuntimeLayout runtimeLayout = new RuntimeLayout(layout, runtime);
+        // If discoverLayoutServers is set, fetch layouts from all other layout servers
+        // discovered in the first search which are not contained in the initial list.
+        if (discoverLayoutServers) {
+            List<String> discoveredLayoutServers = layoutMap.values().stream()
+                    .map(l -> l.getLayoutServers())
+                    .flatMap(s -> s.stream())
+                    .distinct()
+                    .filter(s -> !layoutServers.contains(s))
+                    .collect(Collectors.toList());
+
+            if (!discoveredLayoutServers.isEmpty()) {
+                Map <String, Layout> recursiveLayouts = getLayouts(new ArrayList<>(discoveredLayoutServers), false);
+                layoutMap.putAll(recursiveLayouts);
+            }
+        }
 
-        // All configured nodes.
-        Set<String> allNodes = new HashSet<>(layout.getAllServers());
+        return layoutMap;
+    }
 
-        // Counters to track heartbeat responses from nodes.
-        // A counter is incremented even if we get a WrongEpochException as the node is alive.
-        Map<String, Integer> counters = new HashMap<>();
-        allNodes.forEach(endpoint -> counters.put(endpoint, 0));
+    private Layout getLayoutFromQuorum(Map<String, Layout> layoutMap, int quorum) {
+        Layout layout = null;
+        Map<Layout, Integer> uniqueLayoutMap = null;
+
+        // Find if layouts are the same in all nodes (at least quorum nodes should agree in the
+        // same layout) - count occurrences of each layout in cluster nodes.
+        for (Entry<String, Layout> entry : layoutMap.entrySet()) {
+            if (uniqueLayoutMap == null) {
+                uniqueLayoutMap = new HashMap<Layout, Integer>() {{
+                    put(entry.getValue(), 1);
```

Comment:
Findbugs won't be happy. You create (compiler creates) an anonymous class, which is not so good. You can use ImmutableMap.of() or just create and then put the value

Reply:
I won't address this as this is the part of the code likely to be simplified by the read quorum API.